aws/resource_aws_acm_certificate.go

```diff
@@ -57,16 +57,15 @@ func resourceAwsAcmCertificate() *schema.Resource {
 				ForceNew: true,
 			},
 			"domain_name": {
+				// AWS Provider 3.0.0 aws_route53_zone references no longer contain a
+				// trailing period, yet to account for custom user input, a StateFunc
+				// is in place to prevent ACM API error
 				Type:          schema.TypeString,
 				Optional:      true,
 				Computed:      true,
 				ForceNew:      true,
 				ConflictsWith: []string{"private_key", "certificate_body", "certificate_chain"},
-				StateFunc: func(v interface{}) string {
-					// AWS Provider 1.42.0+ aws_route53_zone references may contain a
-					// trailing period, which generates an ACM API error
-					return strings.TrimSuffix(v.(string), ".")
-				},
+				StateFunc:     trimTrailingPeriod,
```

Comment:
If the ACM API does not accept trailing periods, which if I remember correctly does not, I think our goal during this major release should be fixing up the resource to match the API expectations and return a validation error. Since we are fixing the source of the generally unexpected trailing periods by removing them from the Route 53 data sources/resources, my recommendation would be to remove the special handling like this and the custom resource logic from downstream attribute handling. I'm worried about future upkeep and harder code generation. 👍  We can add a `ValidateFunc` to catch trailing periods to help catch these sooner.